### PR TITLE
Incorporate new binding table.

### PR DESCRIPTION
**Pull request by andywingo**
This is a big ol' commit!  It does many things at once, and in fact
un-does some things.

Firstly, it switches the binding table to use the new parser.  It also
changes the syntax of the binding table.  See the README for
compile_binding_table for more details.  It re-incoporates the address
map into the binding table, because we don't want to run the risk that
there is somehow skew between the binding table at time T and the
address table at time T' -- the need to be all compiled together and,
when loaded in compiled form, all loaded together.

The binding table is compiled automatically and as needed.  We can also
compile the binding table ahead of time, using the compile-binding-table
command (which replaces compile-address-map).

The default name for the binding table in our configs is now
binding-table.txt.  By default compilation will strip a .txt suffix and
replace it with .o.

The binding table has been integrated into the lwaftr, but there are
things that can be cleaned up there.  In particular the clownshoes about
whether an IPv4 address is a big-endian integer, a little-endian
integer, or a uint8_t[4] is a bit irritating.  Dunno.  I'm feeling
pretty good about this!

We also need to change the PSID info map (previously called the address
map) to return 0 for the PSID parameters when an IPv4 address is not in
the map.  This will allow for cheap hairpinning checks instead of doing
a full binding table lookup.

Performance-wise we seem to have a regression, at least from what I can
tell locally -- going down from 18/16 Gbps in the benchmark result to
something more like 17/15 Gbps.  I hope that we can speed this up,
firstly with this address map change, and thereafter with other fixes
like removing C.foo calls from hot paths.

### CODE REVIEW

**Comment by andywingo** [comment]
Pushed a rebase on top of olive.  The previous patch had squashed in a couple patches that were already merged, so the rebase makes the diff a bit smaller maybe.

File: src/apps/lwaftr/binding_table.lua [hunk 1]
```diff
@@ -1,106 +1,469 @@
...
+   if res then return self.softwires:val_at(res) end
+   return nil
+end
+
+function BindingTable:lookup_psid(ipv4, port)
+   local psid_info = self.psid_map:lookup(ipv4).value
+   local psid_len, shift = psid_info.psid_length, psid_info.shift
+   local psid_mask = lshift(1, psid_len) - 1
+   local psid = band(rshift(port, shift), psid_mask)
+   -- Are there are restricted ports for this address?
```
Comment (dpino) [inline src/apps/lwaftr/binding_table.lua]: Nit: "Are there any..."

File: src/apps/lwaftr/binding_table.lua [hunk 2]
```diff
@@ -1,106 +1,469 @@
...
+   if psid_len + shift < 16 then
+      local reserved_ports_bit_count = 16 - psid_len - shift
+      local first_allocated_port = lshift(1, reserved_ports_bit_count)
+      -- The port is within the range of restricted ports.  Assign a
```
Comment (dpino) [inline src/apps/lwaftr/binding_table.lua]: Nit: there are two-spaces after period.

File: src/apps/lwaftr/binding_table.lua [hunk 3]
```diff
@@ -1,106 +1,469 @@
...
+   while not parser:check('}') do
+      local entry = parser:parse_property_list(softwire_spec, '{', '}')
+      key.ipv4, key.psid = entry.ipv4, entry.psid
+      value.br, value.b4_ipv6 = entry.aftr, entry.b4
+      map:add(key, value)
+      local success = true
+      --local success = pcall(map.add, map, key, value)
+      if not success then
+         parser:error('duplicate softwire for ipv4=%s, psid=%d',
+                      lwdebug.format_ipv4(key.ipv4), key.psid)
```
Comment (dpino) [inline src/apps/lwaftr/binding_table.lua]: It seems like originally `map.add` can notify whether an entry was added successfully or not, but the line is commented and `success` is always true, so the error is never going to be reported. If that's what is intended, I would remove the commented line and add a TODO comment. How does it sound?

File: src/apps/lwaftr/lwaftr.lua [hunk 4]
```diff
@@ -194,24 +180,15 @@ local function get_lwAFTR_ipv6(lwstate, binding_entry)
...
-      local bind = lwstate.binding_table[i]
-      if debug then print("CHECK", string.format("%x, %x", bind[2], ipv4_ip)) end
-      if bind[2] == ipv4_ip then
-         local psid_info = bind[3]
-         if psid_info.psid == psid then
-            local lwaftr_ipv6 = get_lwAFTR_ipv6(lwstate, bind)
-            return bind[1], lwaftr_ipv6
-         end
-      end
+   local host_endian_ipv4 = C.htonl(ipv4_ip)
```
Comment (dpino) [inline src/apps/lwaftr/lwaftr.lua]: Shouldn't be `C.ntohl` ?

File: src/apps/lwaftr/podhashmap.lua [hunk 5]
```diff
@@ -75,58 +75,38 @@ function PodHashMap.new(key_type, value_type, hash_fn)
...
+   local map = PodHashMap.new(key_t, value_t, hash_fn)
+   local header = stream:read_ptr(header_t)
+   assert(header.entry_size == header.entry_size)
+   map.size = header.size
+   map.scale = map.size / HASH_MAX
+   map.occupancy = header.occupancy
+   map.max_displacement = header.max_displacement
+   map.entries = stream:read_array(map.entry_type,
+                                   map.size + map.max_displacement + 1)
+   -- ?
```
Comment (dpino) [inline src/apps/lwaftr/podhashmap.lua]: Nit: I don't know what this comment means? :)

File: src/apps/lwaftr/stream.lua [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+   mem = ffi.cast("uint8_t*", mem)
+   local pos = 0
+   local ret = {
+      name=filename,
+      mtime_sec=stat.st_mtime,
+      mtime_nsec=stat.st_mtime_nsec
+   }
+   function ret:close()
+      -- S.munmap(mem, size)
+      mem, pos = nil, nil
```
Comment (dpino) [inline src/apps/lwaftr/stream.lua]: So it's not necessary to unmap `mem`? If assigning nil is enough, maybe commented code can be removed.

File: src/program/snabb_lwaftr/bench/bench.lua [hunk 7]
```diff
@@ -45,5 +45,5 @@ function run(args)
    config.link(c, "statisticsv6.output -> sinkv6.input")
 
    app.configure(c)
-   app.main({})
+   app.main({duration=10})
```
Comment (dpino) [inline src/program/snabb_lwaftr/bench/bench.lua]: Nit: Ideally I would prefer to have a duration flag, like other scripts, with a default to 10.

**Comment by dpino** [comment]
LGTM.

**Comment by andywingo** [comment]
Fixed nits.  The only one not fixed is the munmap one; we really do leak memory.  Unfortunately we'll need to do something complicated though, as memory in the backing store is mapped to all three subtables.  For now let's save ourselves the trouble of spurious segfaults while we're integrating this thing :)

**Pull request merged**